packages/components/src/Notes.js

```diff
@@ -1,15 +1,31 @@
-import React, { useEffect } from 'react'
+import React, { useEffect, useRef } from 'react'
 import { useDeck } from './context'
+import { useStep } from './useStep'
 
 export const Notes = props => {
   const context = useDeck()
+  if (!context || !context.register) {
+    // The slide was rendered as a preview by the <Presenter> component.
+    return false
+  }
+
+  const index = useRef(context.index).current
+  const step = useStep()
   useEffect(() => {
-    if (!context || !context.register) return
-    if (typeof context.index === 'undefined') return
-    context.register(context.index, {
+    // The context changes before this component is unmounted.
+    if (context.index !== index) return
+    // Only render notes for the current step.
+    if (context.step !== step) return
+
+    // Only re-render if the notes changed.
+    const meta = context.metadata[index] || {}
+    if (meta.notes === props.children) return
```

Comment:
Avoid constantly re-rendering the `<MDXDeck>` component by skipping the `context.register` call when the notes have not changed.